### ISSUE TITLE
Wrong deletion during PoseGraph3D cleanup.

### ISSUE THREAD

**Issue by gaschler** [googlecartographer/cartographer#1038]
How to reproduce:
Compile with `CXXFLAGS='-fno-omit-frame-pointer -fsanitize=address -g'` and clang.
Run with ASAN.
```
ASAN_SYMBOLIZER_PATH=/usr/lib/llvm-3.8/bin/llvm-symbolizer \
  ASAN_OPTIONS="detect_leaks=1 detect_stack_use_after_return=true check_initialization_order=true" \
  ./cartographer.mapping.map_builder_test --gtest_filter=*LocalSlam3D
```

In the cleanup phase of the 3D integration test, there is a wrong free:
```
=================================================================
==178005==ERROR: AddressSanitizer: attempting free on address which was not malloc()-ed: 0x60600001b000 in thread T0
    #0 0xcdff40 in operator delete(void*) (workspace/catkin_ws/asan/cartographer/cartographer.mapping.map_builder_test+0xcdff40)
    #1 0x10e169f in void std::_Destroy_aux<false>::__destroy<cartographer::mapping::scan_matching::PrecomputationGrid2D*>(cartographer::mapping::scan_matching::PrecomputationGrid2D*, cartographer::mapping::scan_matching::PrecomputationGrid2D*) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/stl_construct.h:108:6
    #2 0x10e099d in std::vector<cartographer::mapping::scan_matching::PrecomputationGrid2D, std::allocator<cartographer::mapping::scan_matching::PrecomputationGrid2D> >::~vector() /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/stl_vector.h:434:9
    #3 0x10e37a5 in std::default_delete<cartographer::mapping::scan_matching::PrecomputationGridStack>::operator()(cartographer::mapping::scan_matching::PrecomputationGridStack*) const /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/unique_ptr.h:78:2
    #4 0x10ddec0 in std::unique_ptr<cartographer::mapping::scan_matching::PrecomputationGridStack, std::default_delete<cartographer::mapping::scan_matching::PrecomputationGridStack> >::~unique_ptr() /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/unique_ptr.h:268:4
    #5 0x10f2a8a in cartographer::mapping::scan_matching::FastCorrelativeScanMatcher3D::~FastCorrelativeScanMatcher3D() workspace/catkin_ws/src/cartographer/cartographer/mapping/internal/3d/scan_matching/fast_correlative_scan_matcher_3d.cc:157:64
    #6 0x10ee9b5 in std::default_delete<cartographer::mapping::scan_matching::FastCorrelativeScanMatcher3D>::operator()(cartographer::mapping::scan_matching::FastCorrelativeScanMatcher3D*) const /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/unique_ptr.h:78:2
    #7 0x10ede20 in std::unique_ptr<cartographer::mapping::scan_matching::FastCorrelativeScanMatcher3D, std::default_delete<cartographer::mapping::scan_matching::FastCorrelativeScanMatcher3D> >::~unique_ptr() /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/unique_ptr.h:268:4
    #8 0x10ee4f1 in std::_Rb_tree<cartographer::mapping::SubmapId, std::pair<cartographer::mapping::SubmapId const, cartographer::mapping::pose_graph::ConstraintBuilder3D::SubmapScanMatcher>, std::_Select1st<std::pair<cartographer::mapping::SubmapId const, cartographer::mapping::pose_graph::ConstraintBuilder3D::SubmapScanMatcher> >, std::less<cartographer::mapping::SubmapId>, std::allocator<std::pair<cartographer::mapping::SubmapId const, cartographer::mapping::pose_graph::ConstraintBuilder3D::SubmapScanMatcher> > >::_M_drop_node(std::_Rb_tree_node<std::pair<cartographer::mapping::SubmapId const, cartographer::mapping::pose_graph::ConstraintBuilder3D::SubmapScanMatcher> >*) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/stl_tree.h:658:2
    #9 0x10ee428 in std::_Rb_tree<cartographer::mapping::SubmapId, std::pair<cartographer::mapping::SubmapId const, cartographer::mapping::pose_graph::ConstraintBuilder3D::SubmapScanMatcher>, std::_Select1st<std::pair<cartographer::mapping::SubmapId const, cartographer::mapping::pose_graph::ConstraintBuilder3D::SubmapScanMatcher> >, std::less<cartographer::mapping::SubmapId>, std::allocator<std::pair<cartographer::mapping::SubmapId const, cartographer::mapping::pose_graph::ConstraintBuilder3D::SubmapScanMatcher> > >::_M_erase(std::_Rb_tree_node<std::pair<cartographer::mapping::SubmapId const, cartographer::mapping::pose_graph::ConstraintBuilder3D::SubmapScanMatcher> >*) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/stl_tree.h:1858:4
    #10 0x10e7372 in cartographer::mapping::pose_graph::ConstraintBuilder3D::~ConstraintBuilder3D() workspace/catkin_ws/src/cartographer/cartographer/mapping/internal/3d/pose_graph/constraint_builder_3d.cc:72:1
    #11 0xf648f3 in cartographer::mapping::PoseGraph3D::~PoseGraph3D() workspace/catkin_ws/src/cartographer/cartographer/mapping/internal/3d/pose_graph_3d.cc:54:1
    #12 0xf64ecd in cartographer::mapping::PoseGraph3D::~PoseGraph3D() workspace/catkin_ws/src/cartographer/cartographer/mapping/internal/3d/pose_graph_3d.cc:50:29
    #13 0xd23f10 in std::unique_ptr<cartographer::mapping::PoseGraph, std::default_delete<cartographer::mapping::PoseGraph> >::~unique_ptr() /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/unique_ptr.h:268:4
    #14 0xd29ba6 in cartographer::mapping::MapBuilder::~MapBuilder() workspace/catkin_ws/src/cartographer/cartographer/mapping/map_builder.h:40:27
    #15 0xd29bdd in cartographer::mapping::MapBuilder::~MapBuilder() workspace/catkin_ws/src/cartographer/cartographer/mapping/map_builder.h:40:26

0x60600001b000 is located 0 bytes to the right of 64-byte region [0x60600001afc0,0x60600001b000)
allocated by thread T1 here:
    #0 0xcdf940 in operator new(unsigned long) (workspace/catkin_ws/asan/cartographer/cartographer.mapping.map_builder_test+0xcdf940)
    #1 0x110d568 in __gnu_cxx::new_allocator<std::unique_ptr<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3>, std::default_delete<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3> > > >::allocate(unsigned long, void const*) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/ext/new_allocator.h:111:27
    #2 0x110d54a in std::allocator_traits<std::allocator<std::unique_ptr<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3>, std::default_delete<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3> > > > >::allocate(std::allocator<std::unique_ptr<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3>, std::default_delete<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3> > > >&, unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/alloc_traits.h:436:16
    #3 0x110d4c1 in std::_Vector_base<std::unique_ptr<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3>, std::default_delete<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3> > >, std::allocator<std::unique_ptr<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3>, std::default_delete<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3> > > > >::_M_create_storage(unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/stl_vector.h:187:27
    #4 0x110d3c1 in std::vector<std::unique_ptr<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3>, std::default_delete<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3> > >, std::allocator<std::unique_ptr<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3>, std::default_delete<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3> > > > >::vector(unsigned long, std::allocator<std::unique_ptr<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3>, std::default_delete<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3> > > > const&) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/stl_vector.h:284:9
    #5 0x110d342 in cartographer::mapping::DynamicGrid<cartographer::mapping::NestedGrid<cartographer::mapping::FlatGrid<unsigned char, 3>, 3> >::DynamicGrid() workspace/catkin_ws/src/cartographer/cartographer/mapping/3d/hybrid_grid.h:255:29
    #6 0x110d252 in cartographer::mapping::HybridGridBase<unsigned char>::HybridGridBase(float) workspace/catkin_ws/src/cartographer/cartographer/mapping/3d/hybrid_grid.h:421:12
    #7 0x110b782 in cartographer::mapping::scan_matching::ConvertToPrecomputationGrid(cartographer::mapping::HybridGrid const&) workspace/catkin_ws/src/cartographer/cartographer/mapping/internal/3d/scan_matching/precomputation_grid_3d.cc:51:24
    #8 0x10fbd20 in cartographer::mapping::scan_matching::PrecomputationGridStack::PrecomputationGridStack(cartographer::mapping::HybridGrid const&, cartographer::mapping::scan_matching::proto::FastCorrelativeScanMatcherOptions3D const&) workspace/catkin_ws/src/cartographer/cartographer/mapping/internal/3d/scan_matching/fast_correlative_scan_matcher_3d.cc:66:37
    #9 0x10f71f1 in cartographer::common::_Unique_if<cartographer::mapping::scan_matching::PrecomputationGridStack>::_Single_object cartographer::common::make_unique<cartographer::mapping::scan_matching::PrecomputationGridStack, cartographer::mapping::HybridGrid const&, cartographer::mapping::scan_matching::proto::FastCorrelativeScanMatcherOptions3D const&>(cartographer::mapping::HybridGrid const&, cartographer::mapping::scan_matching::proto::FastCorrelativeScanMatcherOptions3D const&) workspace/catkin_ws/src/cartographer/cartographer/common/make_unique.h:47:33
    #10 0x10f2691 in cartographer::mapping::scan_matching::FastCorrelativeScanMatcher3D::FastCorrelativeScanMatcher3D(cartographer::mapping::HybridGrid const&, cartographer::mapping::HybridGrid const*, std::vector<cartographer::mapping::TrajectoryNode, std::allocator<cartographer::mapping::TrajectoryNode> > const&, cartographer::mapping::scan_matching::proto::FastCorrelativeScanMatcherOptions3D const&) workspace/catkin_ws/src/cartographer/cartographer/mapping/internal/3d/scan_matching/fast_correlative_scan_matcher_3d.cc:153:11
    #11 0x10edcfd in cartographer::common::_Unique_if<cartographer::mapping::scan_matching::FastCorrelativeScanMatcher3D>::_Single_object cartographer::common::make_unique<cartographer::mapping::scan_matching::FastCorrelativeScanMatcher3D, cartographer::mapping::HybridGrid const&, cartographer::mapping::HybridGrid const*, std::vector<cartographer::mapping::TrajectoryNode, std::allocator<cartographer::mapping::TrajectoryNode> > const&, cartographer::mapping::scan_matching::proto::FastCorrelativeScanMatcherOptions3D const&>(cartographer::mapping::HybridGrid const&, cartographer::mapping::HybridGrid const*&&, std::vector<cartographer::mapping::TrajectoryNode, std::allocator<cartographer::mapping::TrajectoryNode> > const&, cartographer::mapping::scan_matching::proto::FastCorrelativeScanMatcherOptions3D const&) workspace/catkin_ws/src/cartographer/cartographer/common/make_unique.h:47:33
    #12 0x10e8cea in cartographer::mapping::pose_graph::ConstraintBuilder3D::ConstructSubmapScanMatcher(cartographer::mapping::SubmapId const&, std::vector<cartographer::mapping::TrajectoryNode, std::allocator<cartographer::mapping::TrajectoryNode> > const&, cartographer::mapping::Submap3D const*) workspace/catkin_ws/src/cartographer/cartographer/mapping/internal/3d/pose_graph/constraint_builder_3d.cc:161:7
    #13 0xddfa6a in cartographer::common::ThreadPool::DoWork() workspace/catkin_ws/src/cartographer/cartographer/common/thread_pool.cc:76:5
    #14 0x7eff84c267e2  (/usr/lib/x86_64-linux-gnu/libstdc++.so.6+0xbc7e2)

Thread T1 created by T0 here:
    #0 0xc23109 in __interceptor_pthread_create (workspace/catkin_ws/asan/cartographer/cartographer.mapping.map_builder_test+0xc23109)
    #1 0x7eff84c26a98 in std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) (/usr/lib/x86_64-linux-gnu/libstdc++.so.6+0xbca98)
    #2 0xddfda1 in void std::vector<std::thread, std::allocator<std::thread> >::_M_realloc_insert<cartographer::common::ThreadPool::ThreadPool(int)::$_0>(__gnu_cxx::__normal_iterator<std::thread*, std::vector<std::thread, std::allocator<std::thread> > >, cartographer::common::ThreadPool::ThreadPool(int)::$_0&&) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/vector.tcc:415:4
    #3 0xddf198 in cartographer::common::ThreadPool::ThreadPool(int) workspace/catkin_ws/src/cartographer/cartographer/common/thread_pool.cc:32:5
    #4 0xd1d92f in cartographer::mapping::MapBuilder::MapBuilder(cartographer::mapping::proto::MapBuilderOptions const&) workspace/catkin_ws/src/cartographer/cartographer/mapping/map_builder.cc:64:26
    #5 0xce7d0f in cartographer::common::_Unique_if<cartographer::mapping::MapBuilder>::_Single_object cartographer::common::make_unique<cartographer::mapping::MapBuilder, cartographer::mapping::proto::MapBuilderOptions&>(cartographer::mapping::proto::MapBuilderOptions&) workspace/catkin_ws/src/cartographer/cartographer/common/make_unique.h:47:33
    #6 0xce383d in cartographer::mapping::(anonymous namespace)::MapBuilderTest::BuildMapBuilder() workspace/catkin_ws/src/cartographer/cartographer/mapping/map_builder_test.cc:62:20
    #7 0xce55fa in cartographer::mapping::(anonymous namespace)::MapBuilderTest_LocalSlam3D_Test::TestBody() workspace/catkin_ws/src/cartographer/cartographer/mapping/map_builder_test.cc:151:3

SUMMARY: AddressSanitizer: bad-free (workspace/catkin_ws/asan/cartographer/cartographer.mapping.map_builder_test+0xcdff40) in operator delete(void*)
==178005==ABORTING

```

**Comment by gaschler** [comment]
It appears like FastCorrelativeScanMatcher3D tries to delete a PrecomputationGrid2D through a ambiguous, forward-declared PrecomputationGridStack.

**Comment by gaschler** [comment]
Closing as fixed by #1040.
